workflow/executor/executor.go

```diff
@@ -966,95 +906,20 @@ func chmod(artPath string, mode int32, recurse bool) error {
 	return nil
 }
 
-// containerID is a convenience function to strip the 'docker://', 'containerd://' from k8s ContainerID string
-func containerID(ctrID string) string {
-	schemeIndex := strings.Index(ctrID, "://")
-	if schemeIndex == -1 {
-		return ctrID
-	}
-	return ctrID[schemeIndex+3:]
-}
-
 // Wait is the sidecar container logic which waits for the main container to complete.
 // Also monitors for updates in the pod annotations which may change (e.g. terminate)
 // Upon completion, kills any sidecars after it finishes.
 func (we *WorkflowExecutor) Wait(ctx context.Context) error {
-	err := we.RuntimeExecutor.WaitInit()
-	if err != nil {
-		return err
-	}
-	log.Infof("Waiting on main container")
-	mainContainerID, err := we.waitMainContainerStart(ctx)
-	if err != nil {
-		return err
-	}
-	log.Infof("main container started with container ID: %s", mainContainerID)
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
-
 	annotationUpdatesCh := we.monitorAnnotations(ctx)
-	go we.monitorDeadline(ctx, annotationUpdatesCh)
-
-	err = waitutil.Backoff(ExecutorRetry, func() (bool, error) {
-		err := we.RuntimeExecutor.Wait(ctx, mainContainerID)
-		return err == nil, err
-	})
+	go we.monitorDeadline(ctx, common.MainContainerName, annotationUpdatesCh)
+	err := we.RuntimeExecutor.Wait(ctx)
```

Comment:
I've removed retry here, need to check each executor has adequate retry itself

Reply:
Could you elaborate why this is removed?

Reply:
because it is now the responsibility of the `Wait` implementation - the approach may be different in each implementation - but it does mean I need to go over each one and verify they have backoff when needed

Reply:
I've added back the retry, all the executors can tolerant it I think.